data/blog/lost-in-the-haystack.md

```diff
@@ -0,0 +1,295 @@
+
+One of those constraints is cardinality. As it turns out our aggregation key is high-cardinality. We could have millions of unique aggregations in memory at once. This many aggregations means you have to be strategic about the data you aggregate. But our customers don't care about that. They want to see the data they paid to see. This is the central conflict we need to resolve.
+
+### Minimizing Memory Usage
+
+**In the SELECT Clause**
+
+Let's look at an example query.
+
+```sql
+SELECT replay_id, groupArray(url)
+FROM replay_events
+GROUP BY replay_id
+LIMIT 1
+```
+
+The total memory usage of this query is the sum of every unique replay*id plus \_every* URL stored in the database. We absolutely can not hold this in memory. But we need to return the URLs. So, how do we resolve this?
+
+For every request made by a user to the Replay's service we can make two queries to the database. The first query is a preflight. It returns a set of replay-ids after all search and sort conditions are applied. The second query is a data query. It fetches all the data needed to satisfy the user's request using the replay-ids returned by the preflight.
+
+Let's transform our example query into this new format:
+
+```sql
+-- Preflight Query
+SELECT replay_id
+FROM replay_events
+GROUP BY replay_id
+LIMIT 1
+
+-- Data Query
+SELECT replay_id, groupArray(url)
+FROM replay_events
+WHERE replay_id = 'my_replay_id'
+GROUP BY replay_id
+LIMIT 1
+```
+
+The total memory footprint of our preflight query has been reduced by the uncompressed-size of the `url` column. But our data query hasn't increased by that amount. It's only increased by the uncompressed-size of the `url` column _for that replay_id_.
+
+**In the HAVING Clause**
+
+Let's look at another example query:
+
+```sql
+SELECT replay_id
+FROM replay_events
+GROUP BY replay_id
+HAVING has(groupArray(url), 'sentry.io')
+LIMIT 1
+```
+
+We've adopted our changes from the previous step but now, to answer a search condition, we're aggregating the URL column in the HAVING clause. This has identical memory-usage to SELECTing the column directly. So how do you ask the question "is some value contained within the aggregated set" when the aggregated set is too large to fit in memory? You "stream" it!
+
+Instead of aggregating a column and then asking it some question, you can ask the column a question and aggregate its answer before finally asking a question about the aggregated answer! Clear as mud? Let's demonstrate the concept with SQL.
+
+Let's ask the same question again: "does sentry.io exist in the set of aggregated urls". How should we phrase this in SQL? There's the straight-forward approach `has(groupArray(url), 'sentry.io')` and then there's the streaming approach `sum(url = 'sentry.io') > 0`. What we've done here is subtle but has huge implications. Instead of aggregating the url we're aggregating the result of the condition "does this term match this value" which is represented as either a 0 or 1.
```

Comment:
``` suggestion
Let's ask the same question again: "does sentry.io exist in the set of aggregated urls". How should we phrase this in SQL? There's the straightforward approach `has(groupArray(url), 'sentry.io')` and then there's the streaming approach `sum(url = 'sentry.io') > 0`. What we've done here is subtle but has huge implications. Instead of aggregating the url we're aggregating the result of the condition "does this term match this value" which is represented as either a 0 or 1.
```